ktor-server-rate-limiting/src/commonMain/kotlin/io/github/flaxoos/ktor/server/plugins/ratelimiter/implementations/SlidingWindow.kt

```diff
@@ -0,0 +1,130 @@
+
+        updateWeight(now)
+
+        val callSize = callVolumeUnit.callSize(call)
+        if (timestamps.tryAdd(now, callSize)) {
+            return RateLimiterResponse.NotLimited(this).also {
+                logger.debug { "${call.id()}: Passed with size: $callSize" }
+            }
+        }
+
+        val nextTimestampToBeTrimmed =
+            (timestamps.peekNext()?.first ?: error("Queue is empty but call was rejected"))
+        val resetIn = (timeWindowMs - (now - nextTimestampToBeTrimmed)).milliseconds
+        logger.debug { "${call.id()}: Rejected. reset in: $resetIn" }
+        return RateLimiterResponse.LimitedBy(
+            this,
+            resetIn = resetIn,
+            exceededBy = callSize,
+            message = "$capacity calls were already made during $rate"
+        )
+    }
+
+    private fun updateWeight(nowMs: Long) {
+        val cutoff = nowMs - timeWindowMs
+        logger.debug { "Trimming timestamps before: $cutoff" }
+        timestamps.trimWhere { timestamp ->
+            if (timestamp < cutoff) -1 else if (timestamp > cutoff) 1 else 0
+        }
+    }
+}
+
+internal expect fun <T> provideListForQueue(): MutableList<T>
+
+/**
+ * A fixed size queue of weighted entries with safe access
+ */
+class ConcurrentFixedSizeWeightedQueue<T>(
+    /**
+     * the maximum weight, must be greater than 0
+     */
+    private val maxWeight: Int
+) {
+    private val list = provideListForQueue<Pair<T, Double>>()
+    private val size: Int
+        get() = list.size
+    private var weight = 0.0
+    private val lock = reentrantLock()
+
+    override fun toString(): String {
+        return "ConcurrentFixedSizeWeightedQueue(size=$size, weight=$weight, maxWeight=$maxWeight)"
+    }
+
+    init {
+        require(maxWeight > 0) {
+            "maxWeight must be greater than 0"
+        }
+    }
+
+    fun tryAdd(t: T, weight: Double = 1.0) = lock.withLock {
+        if (this.weight + weight <= maxWeight) {
+            this.weight += weight
+            list.add(t to weight).also {
+                logger.debug { "Added $t with weight $weight, current weight is now ${this.weight}" }
+            }
+        } else {
+            false
+        }
+    }
+
+    fun peekNext() = list.firstOrNull()
+
+    fun trimWhere(comparison: (T) -> Int) = lock.withLock {
+        subListWhere(comparison)?.apply {
+            val toTrim = sumOf { it.second }
+            logger.debug { "Trimmed ${this.size} items, weighing: $toTrim from queue with weight $weight" }
+            weight -= toTrim
+            clear()
+        }
+    }
+
+    private fun subListWhere(comparison: (T) -> Int): MutableList<Pair<T, Double>>? {
+        val index = lock.withLock {
+            list.binarySearch {
+                comparison(it.first)
+            }
+        }
+        val insertionPoint = if (index < 0) -index - 1 else index
+        return if (insertionPoint > 0) {
```

Comment:
This can be rewritten

If -> return
return

No ternary

Reply:
why?

Reply:
Simpler

Reply:
It's very idiomatic in kotlin

https://kotlinlang.org/docs/control-flow.html#if-expression